File: lib/tauri/services.ts
import { invoke } from '@tauri-apps/api/core';

export interface AIRequest {
  prompt: string;
  model: string;
  temperature: number;
  maxTokens?: number;
}

export interface AIResponse {
  content: string;
  model: string;
  usage: {
    promptTokens: number;
    completionTokens: number;
    totalTokens: number;
  };
}

export const aiService = {
  /**
   * Set OpenAI API key
   */
  setApiKey: async (apiKey: string): Promise<void> => {
    try {
      await invoke('set_ai_api_key', { apiKey });
    } catch (error) {
      throw new Error(`Failed to set API key: ${error}`);
    }
  },

  /**
   * Check if API key is configured
   */
  hasApiKey: async (): Promise<boolean> => {
    try {
      return await invoke<boolean>('has_ai_api_key');
    } catch (error) {
      throw new Error(`Failed to check API key: ${error}`);
    }
  },

  /**
   * Send request to AI
   */
  sendRequest: async (request: AIRequest): Promise<AIResponse> => {
    try {
      const response = await invoke<AIResponse>('send_ai_request', {
        request: {
          prompt: request.prompt,
          model: request.model,
          temperature: request.temperature,
          max_tokens: request.maxTokens,
        },
      });
      
      return {
        content: response.content,
        model: response.model,
        usage: {
          promptTokens: response.usage.promptTokens || 0,
          completionTokens: response.usage.completionTokens || 0,
          totalTokens: response.usage.totalTokens || 0,
        },
      };
    } catch (error) {
      throw new Error(`AI request failed: ${error}`);
    }
  },

  /**
   * Generate CSS selector from description
   */
  generateSelector: async (
    description: string,
    pageHtml?: string
  ): Promise<string> => {
    try {
      return await invoke<string>('generate_selector', {
        description,
        pageHtml,
      });
    } catch (error) {
      throw new Error(`Failed to generate selector: ${error}`);
    }
  },

  /**
   * Improve existing CSS selector
   */
  improveSelector: async (
    currentSelector: string,
    issue: string
  ): Promise<string> => {
    try {
      return await invoke<string>('improve_selector', {
        currentSelector,
        issue,
      });
    } catch (error) {
      throw new Error(`Failed to improve selector: ${error}`);
    }
  },

  /**
   * Generate workflow from description
   */
  generateWorkflow: async (description: string): Promise<string> => {
    try {
      return await invoke<string>('generate_workflow', { description });
    } catch (error) {
      throw new Error(`Failed to generate workflow: ${error}`);
    }
  },
};

export const storageService = {
  /**
   * Set value in storage
   */
  set: async (key: string, value: string): Promise<void> => {
    try {
      await invoke('storage_set', { key, value });
    } catch (error) {
      throw new Error(`Failed to set storage value: ${error}`);
    }
  },

  /**
   * Get value from storage
   */
  get: async (key: string): Promise<string | null> => {
    try {
      return await invoke<string | null>('storage_get', { key });
    } catch (error) {
      throw new Error(`Failed to get storage value: ${error}`);
    }
  },

  /**
   * Remove value from storage
   */
  remove: async (key: string): Promise<boolean> => {
    try {
      return await invoke<boolean>('storage_remove', { key });
    } catch (error) {
      throw new Error(`Failed to remove storage value: ${error}`);
    }
  },

  /**
   * Clear all storage
   */
  clear: async (): Promise<void> => {
    try {
      await invoke('storage_clear');
    } catch (error) {
      throw new Error(`Failed to clear storage: ${error}`);
    }
  },

  /**
   * Get all storage keys
   */
  keys: async (): Promise<string[]> => {
    try {
      return await invoke<string[]>('storage_keys');
    } catch (error) {
      throw new Error(`Failed to get storage keys: ${error}`);
    }
  },

  /**
   * Check if key exists
   */
  has: async (key: string): Promise<boolean> => {
    try {
      return await invoke<boolean>('storage_has', { key });
    } catch (error) {
      throw new Error(`Failed to check storage key: ${error}`);
    }
  },
};

export const encryptionService = {
  /**
   * Encrypt data with password
   */
  encrypt: async (data: string, password: string): Promise<string> => {
    try {
      return await invoke<string>('encrypt_data', { data, password });
    } catch (error) {
      throw new Error(`Failed to encrypt data: ${error}`);
    }
  },

  /**
   * Decrypt data with password
   */
  decrypt: async (encrypted: string, password: string): Promise<string> => {
    try {
      return await invoke<string>('decrypt_data', { encrypted, password });
    } catch (error) {
      throw new Error(`Failed to decrypt data: ${error}`);
    }
  },

  /**
   * Hash data
   */
  hash: async (data: string): Promise<string> => {
    try {
      return await invoke<string>('hash_data', { data });
    } catch (error) {
      throw new Error(`Failed to hash data: ${error}`);
    }
  },

  /**
   * Generate random string
   */
  generateRandomString: async (length: number): Promise<string> => {
    try {
      return await invoke<string>('generate_random_string', { length });
    } catch (error) {
      throw new Error(`Failed to generate random string: ${error}`);
    }
  },
};

File: src-tauri/src/commands/services.rs
use tauri::State;
use crate::services::{AIService, StorageService, EncryptionService};

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct AIRequest {
    pub prompt: String,
    pub model: String,
    pub temperature: f32,
    pub max_tokens: Option<u32>,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct AIResponse {
    pub content: String,
    pub model: String,
    pub usage: TokenUsage,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct TokenUsage {
    pub prompt_tokens: u32,
    pub completion_tokens: u32,
    pub total_tokens: u32,
}

// AI Service Commands

#[tauri::command]
pub async fn set_ai_api_key(
    ai_service: State<'_, AIService>,
    api_key: String,
) -> Result<(), String> {
    ai_service.set_api_key(api_key);
    Ok(())
}

#[tauri::command]
pub async fn has_ai_api_key(ai_service: State<'_, AIService>) -> Result<bool, String> {
    Ok(ai_service.has_api_key())
}

#[tauri::command]
pub async fn send_ai_request(
    ai_service: State<'_, AIService>,
    request: AIRequest,
) -> Result<AIResponse, String> {
    let ai_request = crate::services::ai_service::AIRequest {
        prompt: request.prompt,
        model: request.model,
        temperature: request.temperature,
        max_tokens: request.max_tokens,
    };

    let response = ai_service.send_request(ai_request).await?;

    Ok(AIResponse {
        content: response.content,
        model: response.model,
        usage: TokenUsage {
            prompt_tokens: response.usage.prompt_tokens,
            completion_tokens: response.usage.completion_tokens,
            total_tokens: response.usage.total_tokens,
        },
    })
}

#[tauri::command]
pub async fn generate_selector(
    ai_service: State<'_, AIService>,
    description: String,
    page_html: Option<String>,
) -> Result<String, String> {
    ai_service.generate_selector(description, page_html).await
}

#[tauri::command]
pub async fn improve_selector(
    ai_service: State<'_, AIService>,
    current_selector: String,
    issue: String,
) -> Result<String, String> {
    ai_service.improve_selector(current_selector, issue).await
}

#[tauri::command]
pub async fn generate_workflow(
    ai_service: State<'_, AIService>,
    description: String,
) -> Result<String, String> {
    ai_service.generate_workflow(description).await
}

// Storage Service Commands

#[tauri::command]
pub async fn storage_set(
    storage: State<'_, StorageService>,
    key: String,
    value: String,
) -> Result<(), String> {
    storage.set(key, value)
}

#[tauri::command]
pub async fn storage_get(
    storage: State<'_, StorageService>,
    key: String,
) -> Result<Option<String>, String> {
    storage.get(&key)
}

#[tauri::command]
pub async fn storage_remove(
    storage: State<'_, StorageService>,
    key: String,
) -> Result<bool, String> {
    storage.remove(&key)
}

#[tauri::command]
pub async fn storage_clear(storage: State<'_, StorageService>) -> Result<(), String> {
    storage.clear()
}

#[tauri::command]
pub async fn storage_keys(storage: State<'_, StorageService>) -> Result<Vec<String>, String> {
    storage.keys()
}

#[tauri::command]
pub async fn storage_has(
    storage: State<'_, StorageService>,
    key: String,
) -> Result<bool, String> {
    storage.has(&key)
}

// Encryption Service Commands

#[tauri::command]
pub async fn encrypt_data(
    encryption: State<'_, EncryptionService>,
    data: String,
    password: String,
) -> Result<String, String> {
    encryption.encrypt(data.as_bytes(), &password)
}

#[tauri::command]
pub async fn decrypt_data(
    encryption: State<'_, EncryptionService>,
    encrypted: String,
    password: String,
) -> Result<String, String> {
    let decrypted = encryption.decrypt(&encrypted, &password)?;
    String::from_utf8(decrypted).map_err(|e| format!("Invalid UTF-8: {}", e))
}

#[tauri::command]
pub async fn hash_data(
    encryption: State<'_, EncryptionService>,
    data: String,
) -> Result<String, String> {
    Ok(encryption.hash(data.as_bytes()))
}

#[tauri::command]
pub async fn generate_random_string(
    encryption: State<'_, EncryptionService>,
    length: usize,
) -> Result<String, String> {
    encryption.generate_random_string(length)
}
